support installation of  luarocks packages as vim plugins

teto opened the pull request:
###### Motivation for this change
neovim has a vibrant lua ecosystem and I would like to encourage vim plugin authors to rely on luaroicks rather than having every user specify the dependencies themselves. May lead nowhere.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

SuperSandro2000 commented on maintainers/scripts/luarocks-packages.csv:
```diff
@@ -78,10 +79,11 @@ mpack,,,,,
 readline,,,,,
 say,,,,,
-std__debug,std._debug,,,,
+std-_debug,std._debug,,,,
```
normalization is not fully working.

teto replied:
what do you mean ? original package name is `std._debug` it gets renamed to `std-_debug`.

SuperSandro2000 commented on maintainers/scripts/update-luarocks-packages:
```diff
@@ -89,6 +89,10 @@ function convert_pkg() {
         return
     fi
+
+	# Normalize package name
+	nix_pkg_name_normalized=$(sed 's/\./-/' <(echo "$nix_pkg_name"))
```
```suggestion
    # Normalize package name
    nix_pkg_name_normalized=$(sed 's/\./-/' <(echo "$nix_pkg_name"))
```

SuperSandro2000 commented on pkgs/misc/vim-plugins/vim-install-hook.sh:
```diff
@@ -0,0 +1,37 @@
+vimPluginInstallPhase() {
+    echo "Executing vimPluginInstallPhase"
+    runHook preInstall
+    # this is an approximation
+        target="$out/$rtpPath/$pname"
+        mkdir -p $out/${rtpPath}
+        cp -r . $target
+
+        # build help tags
+        if [ -d "$target/doc" ]; then
+          echo "Building help tags"
+          if ! @vimBinary@ -N -u NONE -i NONE -n -E -s -V1 -c "helptags $target/doc" +quit!; then
+            echo "Failed to build help tags!"
+            exit 1
+          fi
+        else
+          echo "No docs available"
+        fi
+
+        if [ -n "$addonInfo" ]; then
+          echo "$addonInfo" > $target/addon-info.json
+        fi
+    runHook postInstall
+    echo "Finished executing vimPluginInstallPhase"
+}
```
This is formated inconsistend.

Commit:
vimPlugins.plenary: init from lua package

This commit shows how to convert luarocks packages into (neo)vim ones.
The advantage for neovim lua plugins to register their rockspec (aka
package definition) is that the plugin can express its dependencies and
a few metadata through it.

teto commented:
I am now able to install neovim lua packages registered on luarocks via this. Should be mostly ready.
